extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,43 +14,407 @@
+   * @return {?Array}
+   * @private
+   */
+  getInlineData_() {
+    const scriptElements = childElementsByTag(this.element, 'SCRIPT');
+    if (!scriptElements.length) {
+      return null;
+    }
+    userAssert(scriptElements.length === 1,
+        `${TAG} should contain at most one <script> child`);
+    const scriptElement = scriptElements[0];
+    userAssert(isJsonScriptTag(scriptElement),
+        `${TAG} should be inside a <script> tag with type="application/json"`);
+    const json = tryParseJson(scriptElement.textContent,
+        error => {
+          throw error;
+        });
+    return json['items'] ? json['items'] : [];
+  }
+
+  /**
+   * Creates and returns <div> that contains the template-rendered children.
+   * @return {Element}
+   * @private
+   */
+  createContainer_() {
+    const container = this.element.ownerDocument.createElement('div');
+    container.classList.add('i-amphtml-autocomplete-results');
+    container.setAttribute('role', 'list');
+    toggle(container, false);
+    this.applyFillContent(container, /* replacedContent */ true);
+    return container;
   }
 
   /** @override */
   layoutCallback() {
-    // Actually load your resource or render more expensive resources.
-    return Promise.resolve();
+    // Disable autofill in browsers.
+    this.inputElement_.setAttribute('autocomplete', 'off');
+
+    // No static local data to filter against.
+    if (!this.inlineData_) {
+      return Promise.resolve();
+    }
+
+    // Register event handlers.
+    this.inputElement_.addEventListener('input',
+        this.inputHandler_.bind(this));
+    this.inputElement_.addEventListener('keydown',
+        this.keyDownHandler_.bind(this));
+    this.inputElement_.addEventListener('focus', this.showResults.bind(this));
+    this.inputElement_.addEventListener('blur', this.hideResults.bind(this));
+    this.container_.addEventListener('mousedown',
+        this.selectHandler_.bind(this));
+
+    return this.mutateElement(() => {
+      this.element.classList.add('i-amphtml-autocomplete');
+      this.inputElement_.classList.add('i-amphtml-autocomplete-input');
+      this.renderResults_();
+    });
+  }
+
+  /**
+   * Create and return <div> element from given plan-text item.
+   * @param {string} item
+   * @return {!Element}
+   * @private
+   */
+  createElementFromItem_(item) {
+    const element = this.element.ownerDocument.createElement('div');
+    element.classList.add('i-amphtml-autocomplete-item');
+    element.setAttribute('role', 'listitem');
+    element.textContent = item;
+    return element;
+  }
+
+  /**
+  * Handle rendering results on user input.
+  * @param {Event} event
+  * @return {!Promise}
+  * @private
+  */
+  inputHandler_(event) {
+    if (event.inputType === 'deleteContentBackward') {
+      // Explore options for caching results to avoid repetitive queries.
+    }
+    return this.mutateElement(() => {
+      this.renderResults_();
+    }).then(() => {
+      this.showResults();
```

Comment:
Since `showResults` appends to the DOM, I should have a mutate block inside it because it could be called by the event listener as well as called here. Does anything strange happen when you have nested mutate blocks?

Reply:
@jridgewell to confirm, but my understanding is that measure/mutate blocks should not be nested because it will cause an extra few frames of unnecessary delay for the inner block. You should establish a method contract so that the method either calls its own measure/mutate, or it always expects to be wrapped inside a measure/mutate.

Reply:
I replaced `showResults(): void` and `hideResults(): promise` with a generic `toggleResults(bool) : void` to isolate the toggle and a `toggleResultsHandler_(bool) : Promise` to wrap it in a mutate block. This way it is a more consistent usage pattern and methods can call either depending on what they are doing, if there is an existing `mutateElement` call, etc.

Reply:
nice that's a solid refactor, it looks great